endpoints/s3/src/putobject.cpp

```diff
@@ -85,37 +116,105 @@ class incremental_async_read : public std::enable_shared_from_this<incremental_a
 		beast::http::response<beast::http::empty_body>& _response,
 		std::string _irods_path,
 		bool _upload_part_flag,
+		bool _know_part_offset_flag,
+		size_t _part_offset,
+		std::string _upload_id,
 		std::string _part_filename,
 		std::shared_ptr<irods::experimental::client_connection> _conn)
 		: session_ptr_{_session_ptr->shared_from_this()}
 		, resp_{std::move(_response)}
 		, parser_{_parser}
 		, irods_path_{_irods_path}
 		, upload_part_flag_{_upload_part_flag}
+		, know_part_offset_flag_{_know_part_offset_flag}
+		, part_offset_{_part_offset}
+		, upload_id_{std::move(_upload_id)}
 		, part_filename_{std::move(_part_filename)}
 		, buffer_(irods::s3::get_put_object_buffer_size_in_bytes())
 		, conn_{_conn}
 		, tp_{*conn_}
 	{
+		namespace part_shmem = irods::s3::api::part_shared_memory;
+
 		resp_.version(parser_->get().version());
 		resp_.set("Etag", _irods_path);
 		resp_.keep_alive(parser_->get().keep_alive());
 
-		if (upload_part_flag_) {
-			logging::trace("{}: Opening part file [{}] for writing.", __func__, part_filename_);
+		if (upload_part_flag_ && know_part_offset_flag_) {
+			// we know the offset so seek and stream directly to iRODS
+			{
+				std::lock_guard<std::mutex> guard(part_shmem::part_size_map_mutex);
+
+				// if there is no replica token then just open the object without replica token and save the token
+				if (part_shmem::replica_token_and_number_map.find(upload_id_) ==
+				    part_shmem::replica_token_and_number_map.end()) {
+					logging::debug(
+						"{}: Opening new iRODS file [{}] for writing and seeking to {}.",
+						__func__,
+						irods_path_,
+						part_offset_);
+					odstream_.open(
+						tp_,
+						std::move(irods_path_),
+						irods::experimental::io::root_resource_name{irods::s3::get_resource()},
+						std::ios::out | std::ios::trunc);
+					if (odstream_.is_open()) {
+						part_shmem::replica_token_and_number_map[upload_id_] =
+							std::pair(odstream_.replica_token(), odstream_.replica_number());
+					}
+				}
+				else {
+					// get the replica token and pass it to open
+					logging::debug(
+						"{}: Opening iRODS file [{}] for writing and seeking to {}.",
+						__func__,
+						irods_path_,
+						part_offset_);
+					odstream_.open(
+						tp_,
+						part_shmem::replica_token_and_number_map[upload_id_].first, // replica token
+						std::move(irods_path_),
+						part_shmem::replica_token_and_number_map[upload_id_].second, // replica number
+						std::ios::out | std::ios::in);
+
+					// It is possible that all of the files have been closed so the aforementioned replica token
+					// is no longer valid. If so, open again without truncate and store the replica token again.
+					if (!odstream_.is_open()) {
+						odstream_.open(
+							tp_,
+							std::move(irods_path_),
+							irods::experimental::io::root_resource_name{irods::s3::get_resource()},
+							std::ios::out | std::ios::in);
+					}
+					if (odstream_.is_open()) {
```

Comment:
I updated the code to handle the comment in your first paragraph.

The difference between the two cases is that one is when chunked encoding is used and another is when it is not used.  I think this is noticeable when looking at the larger context.

I would agree that there is repetition and this could be better handled by some type of design pattern, or at least breaking this section into its own method.  Haven't given it a lot of thought though.

Reply:
let's make an issue for any refactoring.   new functionality is the main goal of this PR.

Reply:
For some reason if I move that `is_open()` check up within the conditional block this crashes sometimes.  I have not reproduced the crash when I don't move this.

I am not sure why.  @alanking Do you see anything?  I did remove the previous `std::move()` on `irods_path_` as this variable is still used but that wasn't it.